### PR TITLE
 Test sources should be available via Eclipse project dependency when target project applies java-test-fixtures plugin

### PR DESCRIPTION
**Pull request by donat**
See description at https://github.com/gradle/gradle/issues/14932

The implementation introduces a new `EclipseClasspath.containsTestFixtures` configuration that lets users configure this behavior explicitly.

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java [hunk 1]
```diff
@@ -275,6 +275,13 @@ public void execute(GenerateEclipseClasspath task) {
                 });
             }
         });
+
+        project.getPlugins().withType(JavaTestFixturesPlugin.class, new Action<JavaTestFixturesPlugin>() {
```
Comment (JLLeitschuh) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:279]: Don't you want this instead? (Lazy task evaluation):

```suggestion
        project.getPlugins().withType(JavaTestFixturesPlugin.class).configureEach(new Action<JavaTestFixturesPlugin>() {
```

Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:279]: That might make sense, but it should not cause any trouble as all the other configurations in this class are done the same way (wo/`configureEach`).

**Comment by donat** [comment]
@gradle/build-tool @jjohannes If nobody objects, I'll merge this PR at the end of my day.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java (hunk 1 above)
Comment (jjohannes) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:279]: This is about plugins. Not tasks. It should not matter which one you use. I would keep it as it is to be consisten.

I think it is confusing that the lazy methods were added to all domain object container. Many of them won't need them. The plugin container is one example.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java [hunk 2]
```diff
@@ -20,31 +20,31 @@
...
 
     @Override
     public DisplayName getDisplayName() {
-        return Describables.withTypeAndName("Eclipse project", eclipseProject.getName());
+        return Describables.withTypeAndName("Eclipse project", eclipseProject.getProject().getName());
     }
 
     public String getName() {
-        return eclipseProject.getName();
+        return eclipseProject.getProject().getName();
```
Comment (jjohannes) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java:47]: Why this change?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java [hunk 3]
```diff
@@ -355,4 +361,26 @@ public FileReferenceFactory getFileReferenceFactory() {
...
     }
+
+    /**
+     * Returns {@code true} if the classpath contains test fixture classes that should be visible
+     * through incoming project dependencies.
+     *
+     * @since 6.8
+     */
+    @Incubating
+    public boolean getContainsTestFixtures() {
```
Comment (jjohannes) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java:372]: :x: This should be a getter with a boolean property and there should be no `set` method. 

Maybe a better name for the property would be `includeTestFixtures`? I think that fits how we name APIs in other places better.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java [hunk 4]
```diff
@@ -275,6 +275,13 @@ public void execute(GenerateEclipseClasspath task) {
...
+            @Override
+            public void execute(JavaTestFixturesPlugin javaTestFixturesPlugin) {
+                model.getClasspath().setContainsTestFixtures(true);
```
Comment (jjohannes) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:282]: When this is turned into a property, this should use `convention`.

**Review by jjohannes — CHANGES_REQUESTED** [review]

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java (hunk 2 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java:47]: The type of the field has changed (to give access to `EclipseClasspath`). I've just renamed the field to avoid confusion.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java (hunk 3 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java:372]: I just wanted to keep the file consistent. WDYT what should be the strategy? Are we fine with having some properties and some fields mixed in this class?

Comment (jjohannes) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java:372]: For me it is clear that everything new has to be a property. It's only mixed because we have not migrated the old fields yet. Which we also have to do.

Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java:372]: Converted the new field.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java (hunk 4 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:282]: Fixed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java (hunk 3 above)
Comment (jjohannes) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java:372]: I don't see any change :)

Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseClasspath.java]: Now?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java [hunk 5]
```diff
@@ -56,4 +56,8 @@ public File getFile() {
     public Set<? extends Task> getGeneratorTasks() {
         return Collections.singleton(generatorTask.get());
     }
+
+    public boolean hasJavaTestFixtures() {
+        return eclipseModel.getClasspath().getContainsTestFixtures().getOrElse(false);
```
Comment (jjohannes) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java:61]: I think it would be nicer if we always set a convention and only do `get()` here. Which means adding `containsTestFixtures.convention(false)` in the `EclipseClasspath ` constructor.

You may change it if you agree (or leave it as it is).

**Review by jjohannes — APPROVED** [review]
Thanks for changing this to a real property. I have one remark which you may or may not address.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java (hunk 5 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/internal/EclipseProjectMetadata.java]: Yeah, that makes sense; I've pushed the implementation.

**Review by jjohannes — APPROVED** [review]

**Comment by donat** [comment]
Ready for release build against the `release` branch: https://builds.gradle.org/viewQueued.html?itemId=39591942

**Comment by donat** [comment]
Thanks @jjohannes for the suggestions!
